primitives/phragmen/src/reduce.rs

```diff
@@ -0,0 +1,1065 @@
+//!    n is fractionally assigned to n 's elected neighbors.
+//!  - `E_w := { e ∈ E : w_e > 0 }`.
+//!
+//!  ### Algorithm overview
+//!
+//! > We consider the input edge weight vector `w` as a directed flow over `E_w` , where the flow in
+//! > each edge is directed from the nominator to the validator. We build `w′` from `w` by removing
+//! > **circulations** to cancel out the flow over as many edges as possible, while preserving flow
+//! > demands over all vertices and without reverting the flow direction over any edge. As long as
+//! > there is a cycle, we can remove an additional circulation and eliminate at least one new edge
+//! > from `E_w′` . This shows that the algorithm always progresses and will eventually finish with
+//! > an acyclic edge support. We keep a data structure that represents a forest of rooted trees,
+//! > which is initialized as a collection of singletons – one per vertex – and to which edges in
+//! > `E_w` are added one by one, causing the trees to merge. Whenever a new edge creates a cycle,
+//! > we detect it and destroy it by removing a circulation. We also run a pre-computation which is
+//! > designed to detect and remove cycles of length four exclusively. This pre-computation is
+//! > optional, and if we skip it then the running time becomes `O (|E_w| ⋅ m), so the
+//! > pre-computation makes sense only if `m >> k` and `|E_w| >> m^2`.
+//! >
+//!
+//! ### Resources:
+//!
+//! 1. https://hackmd.io/JOn9x98iS0e0DPWQ87zGWg?view
+
+use crate::node::{Node, NodeId, NodeRef, NodeRole};
+use crate::{ExtendedBalance, IdentifierT, StakedAssignment};
+use sp_runtime::traits::{Bounded, Zero};
+use sp_std::{
+	collections::btree_map::{BTreeMap, Entry::*},
+	prelude::*,
+};
+
+/// Map type used for reduce_4. Can be easily swapped with HashMap.
+type Map<A> = BTreeMap<(A, A), A>;
+
+/// Returns all combinations of size two in the collection `input` with no repetition.
+fn combinations_2<T: Clone>(input: &[T]) -> Vec<(T, T)> {
+	let n = input.len();
+	if n < 2 {
+		return Default::default();
+	}
+
+	let mut comb = Vec::with_capacity(n * (n - 1) / 2);
+	for i in 0..n {
+		for j in i + 1..n {
+			comb.push((input[i].clone(), input[j].clone()))
+		}
+	}
+	comb
+}
+
+/// Returns the count of trailing common elements in two slices.
+pub(crate) fn trailing_common<T: Eq>(t1: &[T], t2: &[T]) -> usize {
+	let mut t1_pointer = t1.len() - 1;
+	let mut t2_pointer = t2.len() - 1;
+	let mut common = 0usize;
+
+	while t1[t1_pointer] == t2[t2_pointer] {
+		common += 1;
+		if t1_pointer == 0 || t2_pointer == 0 {
+			break;
+		}
+		t1_pointer -= 1;
+		t2_pointer -= 1;
+	}
+
+	common
+}
+
+/// Merges two parent roots as described by the reduce algorithm.
+fn merge<A: IdentifierT>(voter_root_path: Vec<NodeRef<A>>, target_root_path: Vec<NodeRef<A>>) {
+	if voter_root_path.len() <= target_root_path.len() {
+		// iterate from last to beginning, skipping the first one. This asserts that
+		// indexing is always correct.
+		voter_root_path
```

Comment:
Both branches are almost identical. You should just do the following:
```
if voter_root_path.len() <= target_root_path.len() {
    (&voter_root_path, &target_root_path)
else {
    (&target_root_path, &voter_root_path)
}
```

Than you can just write the code once.